### PR TITLE
Fund tranfer feature with in own accounts has been added

### PR DESCRIPTION
**Pull request by vjs3**
With this PR user can now transfer their funds to their own saving or loan accounts.

### CODE REVIEW

File: app/src/main/java/org/mifos/selfserviceapp/api/ApiEndPoints.java [hunk 1]
```diff
@@ -12,4 +12,5 @@
     public static final String AUTHENTICATION = "authentication";
     public static final String CLIENTS = "clients";
     public static final String SAVINGSACOUNTS = "savingsaccounts";
+    public static final String ACCOUNTTRANSFERS = "accounttransfers";
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/api/ApiEndPoints.java]: usually the style here is ACCOUNT_TRANSFERS which makes the name easier to read. without the _ it's a little difficult to understand.

File: app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+public class FundTransferTemplateResponse {
+    private List<FundTransferTemplate> fromAccountOptions = new ArrayList<>();
+    private List<FundTransferTemplate> toAccountOptions = new ArrayList<>();
+
+    public List<FundTransferTemplate> getFromAccountOptions() {
+        return fromAccountOptions;
+    }
+
+    public void setFromAccountOptions(List<FundTransferTemplate> fromAccountOptions) {
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java]: nit: you may not need the set methods for these. Since these are usually populated by retrofit, you can just have the get methods here.

File: app/src/main/java/org/mifos/selfserviceapp/presenters/FundTransferPresenter.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    public void submitTransfer(FundTransferRequest fundTransferRequest) {
+        Call<FundTransferResponse> call = dataManager.submitTransfer(fundTransferRequest);
+        getMvpView().showProgress();
+
+        call.enqueue(new Callback<FundTransferResponse>() {
+            @Override
+            public void onResponse(Response<FundTransferResponse> response) {
+                if (response.code() == 200) {
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/presenters/FundTransferPresenter.java]: this logic is present everywhere. Figure out a way to reuse something.. Can be a separate PR with that change.

File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        for (int i = 0; i < fromAccountNumber.size(); i++) {
+            fromAccountItems[i] = fromAccountNumber.get(i).getAccountNo();
+        }
+        ArrayAdapter<String> fromAccountAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, fromAccountItems);
+        spFromAccount.setAdapter(fromAccountAdapter);
+        spFromAccount.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+            @Override
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long
+                    id) {
+                fromClientId = fundTransferTemplate.getFromAccountOptions().
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java]: need to validate the lists for data. Also use the fromAccountNumber list that you have above.

File: app/src/main/java/org/mifos/selfserviceapp/api/ApiEndPoints.java (hunk 1 above)
Comment (vjs3) [inline app/src/main/java/org/mifos/selfserviceapp/api/ApiEndPoints.java]: Completely agree. Mistake at my part. Will change it.

File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        mFundTransferPresenter.attachView(this);
+
+        mFundTransferPresenter.loadAccountTransferTemplate();
+
+        return rootView;
+    }
+
+    @Override
+    public void showFundTransferTemplate(final FundTransferTemplateResponse fundTransferTemplate) {
+        List<FundTransferTemplate> fromAccountNumber = fundTransferTemplate.getFromAccountOptions();
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java]: It's useful to call the variable fromAccountNumberList to make it clear.. otherwise it looks like it's an account number.

File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        for (int i = 0; i < toAccountNumber.size(); i++) {
+            toAccountItems[i] = toAccountNumber.get(i).getAccountNo();
+        }
+        ArrayAdapter<String> toAccountAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, toAccountItems);
+        spToAccount.setAdapter(toAccountAdapter);
+        spToAccount.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+            @Override
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long
+                    id) {
+                toClientId = fundTransferTemplate.getToAccountOptions().
```
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java]: same as above. Validate data else it might be a candidate for runtime exceptions.

File: app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java (hunk 2 above)
Comment (vjs3) [inline app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java]: Yeah. But I thought it goes with the way I am writing code. Does it make
any difference if yes then please tell me.

File: app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java (hunk 4 above)
Comment (vjs3) [inline app/src/main/java/org/mifos/selfserviceapp/ui/fragments/FundTransferFragment.java]: Yes. Validations are missing as I told you earlier. Will add them.

File: app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java (hunk 2 above)
Comment (satyan) [inline app/src/main/java/org/mifos/selfserviceapp/data/FundTransfer/FundTransferTemplateResponse.java]: yeah. it's not something with this PR. In general, you expose methods that are intended to be used. In this case, set method is not intended to be used.

This is something you can keep in mind in general. Everything you expose a public method, think about whether it's needed..

You may choose to fix it in this PR, or tackle them generally in future PRs. Up to you.
